<!DOCTYPE html>
<html lang="en"><head>
<!-- 2025-01-21 Tue 21:02 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>Monads: Explained Simply</title>
<meta name="author" content="Sawyer Powell" />
<meta name="generator" content="Org Mode" />
<link rel="stylesheet" type="text/css" href="css/index.css"/>
<script defer src="./js/fontawesome/all.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/base16/gruvbox-light-medium.min.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js"></script>
<script src="./js/index.js"></script>
<link rel="preconnect" href="https://fonts.googleapis.com"><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin> <link href="https://fonts.googleapis.com/css2?family=JetBrains+Mono:wght@400;500;600;700&family=Merriweather:ital,wght@0,400;0,700;0,900;1,400;1,700;1,900&family=Open+Sans:ital,wght@0,400;0,600;0,700;1,400;1,600;1,700&display=swap" rel="stylesheet"></head>
<body>
<div class="content">
<div id="sidebar" class="sidebar">
<div class="flex items-center mb-5" onclick="location.href='/'">
<img src="./images/horse.svg" class="shadow-none h-12 rounded-none m-0"/>
<span class="block text-2xl text-fg0 font-bold font-mono ms-2">sawyer-p</span>
</div>
<ul>
<li><a href="about.html">About</a></li>
<li><a href="professional-work.html">Professional Work</a></li>
<hr/>
<li><a href="art.html">Programmatic Art</a></li>
<li><a href="deep-learning.html">Deep Learning</a></li>
<hr/>
<li><a href="posts.html">Posts</a></li>
</ul>
</div>
<div class="org-content">
<h1 class="title">Monads: Explained Simply</h1>
<div class="subtitle">a pragmatic approach - image by alariko</div>
<img src="https://cdna.artstation.com/p/assets/images/images/051/609/672/large/alariko-1649012335676-ilustracion-sin-titulo-01-01.jpg?1657724675" class="hero">
<div class="status"><p class="author">Written by <a target="_blank" href="https://www.linkedin.com/in/sawyerhpowell/">Sawyer Powell</a> - 2025-01-21 Tue 21:02</p></div>
<img src="./images/tiger.svg" class="justify-self-center shadow-none h-12 rounded-none m-0 mb-5"/>

<div id="outline-container-org3b144c8" class="outline-2">
<h2 id="org3b144c8">The Problem</h2>
<div class="outline-text-2" id="text-org3b144c8">
<p>
Let's say we have a function in a Python program that calculates the
reciprocal of a number, adjusted by a random offset between 0 and 1.
</p>

<pre><code class="language-python">import random

def recip(x):
  return 1/(x - random.random())
</code></pre>

<p>
This is a very arbitrary function, but the key idea here is that
randomly, without us predicting when, this function can fail. This is
analogous to reading/writing to a file or database, two operations
which can routinely fail unexpectedly. When <code>x - random.random()</code>
is equal to zero, Python will throw a <code>ZeroDivisionError</code>.
</p>

<p>
Let's say in our program there are instances where we chain together
calls to <code>recip</code> like below.
</p>

<pre><code class="language-python">x = 10
output = recip(x)
output = recip(output)
output = recip(output)
output = recip(output)
</code></pre>

<p>
Somewhere in that codeblock, one of those <code>recip</code> calls could
fail. The simplest way of handling this is just to wrap the entire
block in a <code>try/except</code> clause.
</p>

<pre><code class="language-python">x = 10
try:
  output = recip(x)
  output = recip(output)
  output = recip(output)
  output = recip(output)
except ZeroDivisionError:
  print("Tried to divide by zero")
</code></pre>

<p>
This comes with a cost, every time we
work with <code>recip</code> we have to be thinking about handling the
<code>ZeroDivisionError</code> exception. Working with <code>recip</code> requires mental
overhead. So, it would be nice if we could wrap the <code>recip</code> function
in some way such that we wouldn't have to remember to handle that
exception every time we call it.
</p>

<p>
The key idea here is to wrap the <code>recip</code> function in way that it
returns one of two values:
</p>
<ol class="org-ol">
<li>A <b>success</b> value containing the successful calculation of <code>recip</code></li>
<li>A <b>fail</b> value indicating that <code>recip</code> failed.</li>
</ol>

<p>
Here's one way to wrap it:
</p>

<pre><code class="language-python">class SuccessOrFail:
  def succeed(self, success_val):
    self.success_val = success_val
    self.failed = False

    return self

  def fail(self):
    self.success_val = None
    self.failed = True

    return self

def wrap(func):
  def call_func_safely(x)
    try:
      return SuccessOrFail().succeed(func(x))
    except ZeroDivisionError:
      return SuccessOrFail().fail()

  return call_func_safely
</code></pre>

<p>
Using this wrapper we can now create a "safe" version of our <code>recip</code>
function. Let's call it <code>safe_recip</code>. But, we have a new problem. It's
difficult to chain together calls to <code>safe_recip</code> without needing to
add new control flow.
</p>

<pre><code class="language-python">safe_recip = wrap(recip)

x = 10
output = safe_recip(x)
# !! This could cause unhandled behavior, passing a None into recip
output = safe_recip(output.success_val)
</code></pre>

<p>
To fix this, let's rewrite our <code>wrap</code> function to allow us pass a
<code>SuccessOrFail</code> object as the argument, instead of the plain number
<code>x</code>.
</p>

<pre><code class="language-python">def wrap(func):
  def call_func_safely(success_or_fail)
    if (success_or_fail.failed):
      return SuccessOrFail().fail()

    try:
      return SuccessOrFail().succeed(self.func(success_or_fail.success_val))
    except ZeroDivisionError:
      return SuccessOrFail().fail()

  return call_func_safely
</code></pre>

<p>
Now, we can seamlessly chain together calls to <code>safe_recip</code> without
having to worry at all about handling the failure up front!
</p>

<pre><code class="language-python">safe_recip = wrap(recip)

x = SuccessOrFail().succeed(10)

output = safe_recip(x)
output = safe_recip(output)
output = safe_recip(output)
output = safe_recip(output)
output = safe_recip(output)

if (!output.failed):
  print("Final output is " + str(output.success_val))
else:
  print("Tried to divide by zero")
</code></pre>

<p>
Notice the benefit of this approach over an explicit exception
handling. Wrapping the <code>recip</code> function in this safety net
allows us to write our code in a way where we only have to think about
failure <b>when we actually need to</b>. We're not forced into a position
where we have to program defensively.
</p>

<p>
The exception handling approach adds mental overhead, and it adds
control flow. With this approach we can create pipelines of operations
and only deal with errors when it makes sense to. It makes our code
much easier to reason about, especially in more complicated
situations.
</p>
</div>
</div>

<div id="outline-container-org38871ba" class="outline-2">
<h2 id="org38871ba">Framing Our Problem as a Monad</h2>
<div class="outline-text-2" id="text-org38871ba">
<p>
The first step to thinking <b>monadically</b> is to understand that
functions put our program into a set of possible <b>states</b> once they
are run. In the case of the <code>recip</code> function, when it runs it puts our
programs into one of two possible states:
</p>

<ol class="org-ol">
<li>One where it succeeded and it yielded a value to its caller</li>
<li>One where the function failed and threw an exception</li>
</ol>

<p>
The first step to creating a monad is to encapsulate these possible
states into the value produced by the function. In our solution we
created a new type called <code>SuccessOrFail</code>, and wrapped our <code>recip</code>
function such that it accepts and returns <code>SuccessOrFail</code>. Since
<code>recip</code> itself only accepts a <code>number</code> as input, when we wrap it we
handle the case of <code>SuccessOrFail</code> being in the fail state. When
<code>SuccessOrFail</code> is marked as failed, we simply propagate that state in
the return value of <code>safe_recip</code>.
</p>

<p>
So, to summarize we:
</p>

<ol class="org-ol">
<li>Wrap the number that would have originally been the input to
<code>recip</code> into a type that fully captures the states <code>recip</code> can
yield.</li>
<li>Create a new function which wraps <code>recip</code> such that it is
compatible with taking that new type as input, and producing it as
output.</li>
</ol>

<p>
This type that we create is called the <b>monadic type</b>. The
<code>SuccessOrFail.succeeded(x)</code> function allows us to wrap numbers inside
that type. The <code>wrap</code> function makes functions which take numbers
as inputs and outputs compatible with our monadic type. This <code>wrap</code>
function is known as the <b>combinator</b>, also called the <b>bind</b>
function.
</p>

<p>
So, what specifically is the monad? The monad can be thought of as the
pattern of encapsulating state in a type, providing a way to take
values and embed them in that type, and then a way to make functions
that operate on those values compatible with that new type. The
pattern <b>is</b> the monad.
</p>

<p>
To create a monad you:
</p>
<ul class="org-ul">
<li>Create a new type</li>
<li>Create a function to embed values in that type</li>
<li>Create a function which makes other functions compatible with that
type</li>
</ul>
</div>
</div>

<div id="outline-container-org42b0b19" class="outline-2">
<h2 id="org42b0b19">Popular Uses of Monads</h2>
<div class="outline-text-2" id="text-org42b0b19">
<p>
Monads are excellent choices for handling functions which
fail. Another popular monad is known as the <code>Maybe</code> monad, which helps
deal with values being possibly null.
</p>

<p>
An interesting use case of monads is in logging. Notice that in our
<code>wrap</code> function we can add arbitrary code, which includes logging. In
this sense we are "decorating" whatever functions we pass through
<code>wrap</code> with extra code. 
</p>

<p>
A simple decoration we can do is to <code>print</code> the output of our function
every time it runs. This might be helpful for debugging.
</p>

<pre><code class="language-python">def wrap(func):
  def call_func_safely(success_or_fail)
    if (success_or_fail.failed):
      return SuccessOrFail().fail()

    try:
      success_val = self.func(success_or_fail.success_val)
      print("Succeeded on input " + str(success_or_fail.success_val) + " yielding " + str(success_val))
      return SuccessOrFail().succeed(success_val)
    except ZeroDivisionError:
      print("Failed. Division by zero.")
      return SuccessOrFail().fail()

  return call_func_safely
</code></pre>

<p>
Even more interesting is we can build up a log <b>inside</b> the monadic
type and then read it any time we want to.
</p>

<pre><code class="language-python">class SuccessOrFail:
  def succeed(self, success_val, log=None):
    self.success_val = success_val
    self.failed = False
    self.log = log

    return self

  def fail(self, log=None):
    self.success_val = None
    self.failed = True
    self.log = log

    return self

def wrap(func):
  def call_func_safely(success_or_fail)
    if (success_or_fail.failed):
      return SuccessOrFail().fail()

    try:
      success_val = self.func(success_or_fail.success_val)
      return SuccessOrFail().succeed(success_val, success_or_fail.log + "\nSucceeded on input " + str(success_or_fail.success_val) + " yielding " + str(success_val))
    except ZeroDivisionError:
      print("Failed. Division by zero.")
      return SuccessOrFail().fail(success_or_fail.log + "\nFailed. Divide by zero")

  return call_func_safely
</code></pre>

<p>
This shows one of the amazing "hidden" features of monads, that you
can build up state across chained operations. And can do this without
introducing bloat to other parts of the program.
</p>

<pre><code class="language-python">safe_recip = wrap(recip)

x = SuccessOrFail().succeed(10)

output = safe_recip(x)
output = safe_recip(output)
output = safe_recip(output)
output = safe_recip(output)
output = safe_recip(output)

print(output.log)
</code></pre>

<pre><code class="language-text">Succeeded on input 10 yielding 0.10644867208448258
Succeeded on input 0.10644867208448258 yielding -0.3266419394727856
...
</code></pre>

<p>
Some of the most notable uses of monads outside of functional
languages like Haskell are in Typescript and Rust with their <code>?</code>
operator.
</p>

<p>
In Typescript, the <code>?</code> operator is used to shortcircuit chained
function calls, or property accesses, to a null/undefined value.
</p>

<pre><code class="language-typescript">let first_users_name = response.data?.users?.[0]?.name;
</code></pre>

<p>
Here we try to get the name of the first user in the response. We set
the <code>first_users_name</code> variable to null if any of those property
accesses yields us a null (or undefined) value. A fun exercise is to
reason about why the <code>?</code> can be considered a <b>combinator</b> or <b>bind</b>
function.
</p>
</div>
</div>
</div>
<div id="table-of-contents" role="doc-toc">
<h2>Table of Contents</h2>
<div id="text-table-of-contents" role="doc-toc">
<ul>
<li><a href="#org3b144c8">The Problem</a></li>
<li><a href="#org38871ba">Framing Our Problem as a Monad</a></li>
<li><a href="#org42b0b19">Popular Uses of Monads</a></li>
</ul>
</div>
</div>
</div>
<div class="fixed bottom-10 flex space-x-3"><div id="menu-btn" class="h-12 w-12 p-2 rounded-full bg-aqua text-fg flex justify-center items-center cursor-pointer shadow-lg fill-white 2xl:hidden"><i class="fa-regular fa-bars-sort fa-xl" style="color:white"></i></div><div id="toc-btn" class="h-12 w-12 p-2 rounded-full bg-blue text-fg flex justify-center items-center cursor-pointer shadow-lg fill-white xl:hidden"><i class="fa-regular fa-list-tree fa-xl" style="color:white"></i></div></div></body>
<script>hljs.highlightAll();</script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script></html>
